src/upgrade.py

```diff
@@ -0,0 +1,219 @@
+
+    def _on_resume_upgrade(self, event: ActionEvent) -> None:
+        """Handle resume upgrade action.
+
+        Continue the upgrade by setting the partition to the next unit.
+        """
+        if not self.charm.unit.is_leader():
+            event.fail("Action only valid on leader unit")
+            return
+        fail_message = "Nothing to resume, upgrade stack unset"
+        if self.upgrade_stack:
+            try:
+                next_partition = self.upgrade_stack[-1]
+                event.set_results({"message": f"Upgrade will resume on unit {next_partition}"})
+                self.charm.k8s_helpers.set_rolling_update_partition(partition=next_partition)
+                return
+            except IndexError:
+                fail_message = "Nothing to resume, empty upgrade stack"
+            except KubernetesClientError:
+                fail_message = "Cannot set rolling update partition"
+        event.fail(fail_message)
+
+    def _on_pebble_ready(self, event) -> None:
+        """Handle pebble ready event.
+
+        Confirm that unit is healthy and set unit completed.
+        """
+        if not self.peer_relation:
+            logger.debug("Peer relation not yet ready on unit. Deferring event")
+            event.defer()
+            return
+
+        if self.peer_relation.data[self.charm.unit].get("state") in ["ready", "upgrading"]:
+            try:
+                for attempt in Retrying(stop=stop_after_attempt(6), wait=wait_fixed(10)):
+                    with attempt:
+                        if self.charm._mysql.is_instance_in_cluster(self.charm.unit_label):
+                            logger.debug(
+                                "Upgraded unit is healthy. Set upgrade state to `completed`"
+                            )
+                            self.set_unit_completed()
+                            # call update status asap
+                            self.charm._on_update_status(None)
+                        else:
+                            logger.debug(
+                                "Instance not yet back in the cluster."
+                                f" Retry {attempt.retry_state.attempt_number}/6"
+                            )
+                            raise Exception
+            except RetryError:
+                logger.debug("Upgraded unit is not healthy")
+                self.set_unit_failed()
+                self.charm.unit.status = BlockedStatus(
+                    "upgrade failed. Check logs for rollback instruction"
+                )
+                self.log_rollback_instructions()
+
+    @override
+    def _on_upgrade_charm(self, _: UpgradeCharmEvent) -> None:
+        """Handle upgrade charm."""
+        # only test versions for the first unit being upgraded
+        if self.charm.unit.name == f"{self.charm.app.name}/{self.highest_ordinal}":
+            try:
+                self._upgrade_supported_check()
```

Comment:
also run ` util.checkForServerUpgrade()` command of mysql-shell here
but with NEW version as target-version
https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-utilities-upgrade.html

Reply:
0382943